From 2d6301176789f35c05d052d2a6299dd4e86afc64 Mon Sep 17 00:00:00 2001
From: Ryan Blue <rblue@cloudera.com>
Date: Wed, 1 Oct 2014 14:14:24 -0700
Subject: [PATCH 33/93] PARQUET-107: Add option to disable summary metadata.

This adds an option to the commitJob phase of the MR OutputCommitter,
parquet.enable.summary-metadata (default true), that can be used to
disable the summary metadata files generated from the footers of all of
the files produced. This enables more control over when those summary
files are produced and makes it possible to rename MR outputs and then
generate the summaries.

Author: Ryan Blue <rblue@cloudera.com>

Closes #68 from rdblue/PARQUET-107-add-summary-metadata-option and squashes the following commits:

261e5e4 [Ryan Blue] PARQUET-107: Add option to disable summary metadata.
---
 .../parquet/hadoop/ParquetOutputCommitter.java     |   26 ++++++++++---------
 .../java/parquet/hadoop/ParquetOutputFormat.java   |    5 ++++
 2 files changed, 19 insertions(+), 12 deletions(-)

diff --git a/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputCommitter.java b/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputCommitter.java
index 31917d2..b22bd87 100644
--- a/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputCommitter.java
+++ b/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputCommitter.java
@@ -42,22 +42,24 @@ public class ParquetOutputCommitter extends FileOutputCommitter {
 
   public void commitJob(JobContext jobContext) throws IOException {
     super.commitJob(jobContext);
-    try {
-      Configuration configuration = ContextUtil.getConfiguration(jobContext);
-      final FileSystem fileSystem = outputPath.getFileSystem(configuration);
-      FileStatus outputStatus = fileSystem.getFileStatus(outputPath);
-      List<Footer> footers = ParquetFileReader.readAllFootersInParallel(configuration, outputStatus);
+    Configuration configuration = ContextUtil.getConfiguration(jobContext);
+    if (configuration.getBoolean(ParquetOutputFormat.ENABLE_JOB_SUMMARY, true)) {
       try {
-        ParquetFileWriter.writeMetadataFile(configuration, outputPath, footers);
+        final FileSystem fileSystem = outputPath.getFileSystem(configuration);
+        FileStatus outputStatus = fileSystem.getFileStatus(outputPath);
+        List<Footer> footers = ParquetFileReader.readAllFootersInParallel(configuration, outputStatus);
+        try {
+          ParquetFileWriter.writeMetadataFile(configuration, outputPath, footers);
+        } catch (Exception e) {
+          LOG.warn("could not write summary file for " + outputPath, e);
+          final Path metadataPath = new Path(outputPath, ParquetFileWriter.PARQUET_METADATA_FILE);
+          if (fileSystem.exists(metadataPath)) {
+            fileSystem.delete(metadataPath, true);
+          }
+        }
       } catch (Exception e) {
         LOG.warn("could not write summary file for " + outputPath, e);
-        final Path metadataPath = new Path(outputPath, ParquetFileWriter.PARQUET_METADATA_FILE);
-        if (fileSystem.exists(metadataPath)) {
-          fileSystem.delete(metadataPath, true);
-        }
       }
-    } catch (Exception e) {
-      LOG.warn("could not write summary file for " + outputPath, e);
     }
   }
 
diff --git a/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputFormat.java b/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputFormat.java
index a8e5c0a..0e69be5 100644
--- a/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputFormat.java
+++ b/parquet-hadoop/src/main/java/parquet/hadoop/ParquetOutputFormat.java
@@ -72,6 +72,10 @@ import parquet.hadoop.util.ConfigurationUtil;
  *
  * # To enable/disable dictionary encoding
  * parquet.enable.dictionary=true # false to disable dictionary encoding
+ *
+ * # To enable/disable summary metadata aggregation at the end of a MR job
+ * # The default is true (enabled)
+ * parquet.enable.summary-metadata=true # false to disable summary aggregation
  * </pre>
  *
  * If parquet.compression is not set, the following properties are checked (FileOutputFormat behavior).
@@ -98,6 +102,7 @@ public class ParquetOutputFormat<T> extends FileOutputFormat<Void, T> {
   public static final String ENABLE_DICTIONARY    = "parquet.enable.dictionary";
   public static final String VALIDATION           = "parquet.validation";
   public static final String WRITER_VERSION       = "parquet.writer.version";
+  public static final String ENABLE_JOB_SUMMARY   = "parquet.enable.summary-metadata";
 
   public static void setWriteSupportClass(Job job,  Class<?> writeSupportClass) {
     getConfiguration(job).set(WRITE_SUPPORT_CLASS, writeSupportClass.getName());
-- 
1.7.0.4

